### PR TITLE
New energy manager

### PR DESCRIPTION
**Pull request by corneliusclaussen**
It creates an energy market and each EvseManager now has a broker that buys/sells energy on behalf of the EvseManager. This allows to simply implement new charging strategies.
This commit includes the basic architecture as well as the default FastCharging broker.

Supports:

- AC and DC (also mixed environments)
- Limited bidirectional charging
- schedules for import and export of energy
- prepared for OCPP smart charging compatibility
- enforcing external (schedule) limits at each node of the tree

Known limits:
- It does not use powermeters yet, hence does not wait for one evse to power down before the other powers up
- Bidirectional charging is supported with a hack config option for ISO-2 and DIN SPEC. It needs to be modified for ISO-20.
- Smart charging goals not supported yet (pricing, ...)

### CODE REVIEW

**Review by hikinggrass — CHANGES_REQUESTED** [review]

File: interfaces/energy.yaml [hunk 1]
```diff
@@ -1,97 +1,17 @@
...
-  This interface defines a forecast schedule of energy import/export capability/limits,
-  pricing and current energy usage.
+  This interface is the internal energy management inteface between nodes.
 cmds:
   enforce_limits:
     description: The EnergyManager enforces a limit using this command.
     arguments:
-      uuid:
+      value:
         description: UUID of node that this limit applies to
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply interfaces/energy.yaml:8]: fixed

File: interfaces/energy.yaml [hunk 2]
```diff
@@ -1,97 +1,17 @@
...
-          are absolute UTC time.
-        type: array
-        items:
-          description: One entry for the time series
-          type: object
-          $ref: /energy#/TimeSeriesEntryExtended
-      energy_usage:
-        description: Energy usage of this node (powermeter struct)
-        type: object
+    $ref: /energy#/EnergyFlowRequest
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply interfaces/energy.yaml:17]: No new line at end of ./GenericPowermeter/models/test_dummy.yaml
No new line at end of ./GenericPowermeter/models/template.yaml
No new line at end of ./PacketSniffer/cmake/FindPCAP.cmake

[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply interfaces/energy.yaml:17]: All that relate to this PR have a line ending at the end of file now.

File: interfaces/external_energy_limits.yaml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+description: >-
+  This interface allows to limit energy flow at a specific node of the energy tree from the outside (e.g. from ocpp).
+cmds:
+  set_external_limits:
+    description: Set additional external energy flow limits at this node.
+    arguments:
+      value:
+        description: UUID of node that this limit applies to
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply interfaces/external_energy_limits.yaml:8]: fixed

File: modules/EnergyManager/BrokerFastCharging.cpp [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+        }
+        if (total_power.has_value()) {
+            trading[i].limits_to_root.total_power_W = 0.;
+        }
+    }
+
+    traded = false;
+
+    if (offer->import_offer.size() != offer->export_offer.size())
+        EVLOG_error << "import and export offer do not have the same size!";
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EnergyManager/BrokerFastCharging.cpp:41]: fixed

File: modules/EvseManager/evse/evse_managerImpl.cpp [hunk 5]
```diff
@@ -35,6 +35,10 @@ void evse_managerImpl::init() {
         fmt::format("everest_external/nodered/{}/cmd/set_max_current", mod->config.connector_id),
         [&charger = mod->charger, this](std::string data) { mod->updateLocalMaxCurrentLimit(std::stof(data)); });
 
+    mod->mqtt.subscribe(
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EvseManager/evse/evse_managerImpl.cpp:38]: we should, but let's do it in a seperate pull request as it touches much more then just energy management

File: modules/EvseManager/manifest.yaml [hunk 6]
```diff
@@ -58,7 +58,11 @@ config:
     description: Enable or disable RCD
     type: boolean
     default: true
-  max_current:
+  max_current_import:
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EvseManager/manifest.yaml:61]: fixed

File: modules/JsDCSupplySimulator/index.js
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/JsDCSupplySimulator/index.js:1]: will fix in a seperate commit

File: modules/EvseManager/energy_grid/energyImpl.cpp [hunk 7]
```diff
@@ -7,204 +7,322 @@
...
+                auto tpnow = date::utc_clock::now();
+                auto tp = Everest::Date::to_rfc3339(date::floor<std::chrono::hours>(tpnow) +
+                                                    date::get_leap_second_info(tpnow).elapsed);
+
+                energy_flow_request.schedule_import.value()[0].timestamp = tp;
+                energy_flow_request.schedule_export.value()[0].timestamp = tp;
+
+                // If we need energy, copy local limit schedules to energy_flow_request.
+                if (s == Charger::EvseState::Charging || s == Charger::EvseState::PrepareCharging ||
+                    s == Charger::EvseState::PrepareCharging || s == Charger::EvseState::WaitingForEnergy ||
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EvseManager/energy_grid/energyImpl.cpp]: fixed

File: modules/EvseManager/energy_grid/energyImpl.cpp [hunk 8]
```diff
@@ -7,204 +7,322 @@
...
+                        if (!e.limits_to_root.ac_max_current_A.has_value() ||
+                            e.limits_to_root.ac_max_current_A.value() > hw_caps.max_current_A_import)
+                            e.limits_to_root.ac_max_current_A = hw_caps.max_current_A_import;
+
+                        if (!e.limits_to_root.ac_max_phase_count.has_value() ||
+                            e.limits_to_root.ac_max_phase_count.value() > hw_caps.max_phase_count_import)
+                            e.limits_to_root.ac_max_phase_count = hw_caps.max_phase_count_import;
+
+                        // copy remaining hw limits on root side
+                        e.limits_to_root.ac_min_current_A = hw_caps.max_current_A_import;
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EvseManager/energy_grid/energyImpl.cpp:110]: fixed

File: modules/EvseManager/EvseManager.cpp [hunk 9]
```diff
@@ -167,8 +170,8 @@ void EvseManager::ready() {
                     return;
 
                 // Some power supplies don't really like 0 current limits
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EvseManager/EvseManager.cpp:172]: no, just a left over

File: modules/EnergyManager/Market.cpp [hunk 10]
```diff
@@ -0,0 +1,325 @@
...
+ScheduleRes Market::get_sold_energy() {
+    return sold_root;
+}
+
+Market* Market::parent() {
+    return _parent;
+}
+
+bool Market::is_root() {
+    return _parent == NULL;
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EnergyManager/Market.cpp]: fixed

File: modules/EnergyManager/Market.hpp [hunk 11]
```diff
@@ -0,0 +1,94 @@
...
+private:
+    std::chrono::high_resolution_clock::time_point timepoint_start;
+    std::chrono::nanoseconds total_duration{0};
+    bool running{false};
+};
+
+class Market {
+public:
+    Market(types::energy::EnergyFlowRequest& _energy_flow_request, const float __nominal_ac_voltage,
+           Market* __parent = NULL);
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EnergyManager/Market.hpp]: fixed

File: modules/EnergyManager/EnergyManager.cpp [hunk 12]
```diff
@@ -1,488 +1,138 @@
...
-        obj_to_sanitize["limits_import"] = json::array();
-    }
+    bool trade_happend_in_this_round = true;
 
-    if (obj_to_sanitize.contains("limits_export") &&
-        obj_to_sanitize["limits_export"]
-            .is_null()) { // need "[]" to prevent nlohmann error 304: cannot use at() with null
-        obj_to_sanitize["limits_export"] = json::array();
-    }
+    int timeout = 100;
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EnergyManager/EnergyManager.cpp:79]: fixed

File: modules/EnergyManager/BrokerFastCharging.cpp [hunk 13]
```diff
@@ -0,0 +1,253 @@
...
+        if (globals.debug)
+            EVLOG_info << fmt::format("\033[1;33m                               NO TRADE \033[1;0m");
+
+        //   execute the zero trade on the market
+        local_market.trade(trading);
+        return false;
+    }
+}
+
+bool BrokerFastCharging::buy_ampere_import(int i, float a, bool allow_less) {
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EnergyManager/BrokerFastCharging.cpp:126]: fixed

File: modules/EnergyManager/EnergyManager.cpp [hunk 14]
```diff
@@ -1,488 +1,138 @@
...
-            }
-        }
-
-    } catch (const std::exception& e) {
-        EVLOG_error << "Cannot enforce limits: Exception occurred: r_energy_trunk connection faulty " << e.what();
+void EnergyManager::enforce_limits(const std::vector<types::energy::EnforcedLimits>& limits) {
+    for (const auto& it : limits) {
+        if (globals.debug)
+            EVLOG_info << fmt::format("\033[1;92m{} Enforce limits {}A {}W \033[1;0m", it.uuid,
+                                      it.limits_root_side.value().ac_max_current_A.value_or(-9999),
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/EnergyManager/EnergyManager.cpp:42]: removed 9999 constant, added a proper debug string

File: modules/EnergyManager/BrokerFastCharging.hpp
Comment (hikinggrass) [inline modules/EnergyManager/BrokerFastCharging.hpp:1]: In this header a few "i" and "a" parameter names are still left, fixed already in the cpp file

**Review by hikinggrass — APPROVED** [review]

**Pull request merged**